File: Maximum Length.py
from collections import Counter

class Solution():
    def solve(self, a, b, c):
        prev = -1
        v = [a, b, c]

        ans = ""

        while (1):
            ma = 0
            cur = 0

            for i in range(3):
                if (prev != i + 97 and ma < v[i]):
                    ma = v[i]
                    cur = i + 97

            if (ma == 0):
                break

            ans += chr(cur)
            v[cur - 97]-=1
            if (ma >= 2 and (prev == -1 or ma > v[prev - 97])):
                ans += chr(cur)
                v[cur - 97]-=1
            prev = cur
            n=len(ans)
            
        if n != a + b + c:
            return -1

        return n

File: Maximum Length.cpp
class Solution {
public:
    int solve(int a, int b, int c) {
        // code here
        char prev = '0';
        vector<int> v = {a, b, c};

        string ans;

        while (1) {
            int ma = 0;
            char cur;

            for (int i = 0; i < 3; i++)
                if (prev != char(i + 'a') && ma < v[i])
                    ma = v[i], cur = i + 'a';

            if (ma == 0)
                break;

            ans += cur;
            v[cur - 'a']--;
            if (ma >= 2 && (prev == '0' || ma > v[prev - 'a'])) {
                ans += cur;
                v[cur - 'a']--;
            }
            prev = cur;
        }
        int n=ans.length();
        if(n!=a+b+c) return -1;
        return n;
    }
};
